### ISSUE TITLE
ctx.runfiles has undocumented preconditions

### ISSUE THREAD

**Issue by endobson** [bazelbuild/bazel#3338]
I got the following error
```
illegal argument in call to runfiles: Order mismatch: topological != postorder.
```
When trying to add a depset with topological sorting to runfiles, but nothing in the documentation mentions this requirement.

https://docs.bazel.build/versions/master/skylark/lib/ctx.html#runfiles

Bazel version is 0.5.2

**Comment by endobson** [comment]
The following code generates a similar error but one that crashes bazel.

```
  files2 = depset(order="topological")
  files2 += [ctx.outputs.list]
  files_post = depset(order="postorder")
  files_post += files2
```

```
java.lang.RuntimeException: Unrecoverable error while evaluating node 'CONFIGURED_TARGET://bootstrap:library-compiler-srcs 319ec42f619875d649ef91d6019e9250 (1632251418 1438876033)' (requested by nodes )
        at com.google.devtools.build.skyframe.ParallelEvaluator$Evaluate.run(ParallelEvaluator.java:474)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:357)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: Order mismatch: topological != postorder
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:383)
        at com.google.devtools.build.lib.util.Preconditions.checkArgument(Preconditions.java:90)
        at com.google.devtools.build.lib.collect.nestedset.NestedSetBuilder.addTransitive(NestedSetBuilder.java:117)
        at com.google.devtools.build.lib.syntax.SkylarkNestedSet.<init>(SkylarkNestedSet.java:159)
        at com.google.devtools.build.lib.syntax.SkylarkNestedSet.<init>(SkylarkNestedSet.java:114)
        at com.google.devtools.build.lib.syntax.BinaryOperatorExpression.plus(BinaryOperatorExpression.java:246)
        at com.google.devtools.build.lib.syntax.BinaryOperatorExpression.evaluate(BinaryOperatorExpression.java:125)
        at com.google.devtools.build.lib.syntax.LValue.assign(LValue.java:57)
        at com.google.devtools.build.lib.syntax.AugmentedAssignmentStatement.doExec(AugmentedAssignmentStatement.java:55)
        at com.google.devtools.build.lib.syntax.Statement.exec(Statement.java:31)
        at com.google.devtools.build.lib.syntax.UserDefinedFunction.call(UserDefinedFunction.java:81)
        at com.google.devtools.build.lib.syntax.BaseFunction.call(BaseFunction.java:428)
        at com.google.devtools.build.lib.rules.SkylarkRuleConfiguredTargetBuilder.buildRule(SkylarkRuleConfiguredTargetBuilder.java:87)
        at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createRule(ConfiguredTargetFactory.java:319)
        at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createConfiguredTarget(ConfiguredTargetFactory.java:221)
        at com.google.devtools.build.lib.skyframe.SkyframeBuildView.createConfiguredTarget(SkyframeBuildView.java:489)
        at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.createConfiguredTarget(ConfiguredTargetFunction.java:1136)
        at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.compute(ConfiguredTargetFunction.java:262)
        at com.google.devtools.build.skyframe.ParallelEvaluator$Evaluate.run(ParallelEvaluator.java:400)
        ... 4 more
java.lang.RuntimeException: Unrecoverable error while evaluating node 'CONFIGURED_TARGET://bootstrap:library-compiler-srcs 319ec42f619875d649ef91d6019e9250 (1632251418 1438876033)' (requested by nodes )
        at com.google.devtools.build.skyframe.ParallelEvaluator$Evaluate.run(ParallelEvaluator.java:474)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:357)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: Order mismatch: topological != postorder
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:383)
        at com.google.devtools.build.lib.util.Preconditions.checkArgument(Preconditions.java:90)
        at com.google.devtools.build.lib.collect.nestedset.NestedSetBuilder.addTransitive(NestedSetBuilder.java:117)
        at com.google.devtools.build.lib.syntax.SkylarkNestedSet.<init>(SkylarkNestedSet.java:159)
        at com.google.devtools.build.lib.syntax.SkylarkNestedSet.<init>(SkylarkNestedSet.java:114)
        at com.google.devtools.build.lib.syntax.BinaryOperatorExpression.plus(BinaryOperatorExpression.java:246)
        at com.google.devtools.build.lib.syntax.BinaryOperatorExpression.evaluate(BinaryOperatorExpression.java:125)
        at com.google.devtools.build.lib.syntax.LValue.assign(LValue.java:57)
        at com.google.devtools.build.lib.syntax.AugmentedAssignmentStatement.doExec(AugmentedAssignmentStatement.java:55)
        at com.google.devtools.build.lib.syntax.Statement.exec(Statement.java:31)
        at com.google.devtools.build.lib.syntax.UserDefinedFunction.call(UserDefinedFunction.java:81)
        at com.google.devtools.build.lib.syntax.BaseFunction.call(BaseFunction.java:428)
        at com.google.devtools.build.lib.rules.SkylarkRuleConfiguredTargetBuilder.buildRule(SkylarkRuleConfiguredTargetBuilder.java:87)
        at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createRule(ConfiguredTargetFactory.java:319)
        at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createConfiguredTarget(ConfiguredTargetFactory.java:221)
        at com.google.devtools.build.lib.skyframe.SkyframeBuildView.createConfiguredTarget(SkyframeBuildView.java:489)
        at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.createConfiguredTarget(ConfiguredTargetFunction.java:1136)
        at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.compute(ConfiguredTargetFunction.java:262)
        at com.google.devtools.build.skyframe.ParallelEvaluator$Evaluate.run(ParallelEvaluator.java:400)
        ... 4 more
```

**Comment by brandjon** [comment]
Thanks for the report. For the code sample you gave, the crash has just been fixed to instead be a simple skylark error. Note that for this case the order restriction is documented:

https://docs.bazel.build/versions/master/skylark/lib/depset.html

> Two depsets may only be merged (via + or the union() method) if either both depsets have the same order, or one of them has "default" order.

We'll update the runfiles documentation for the transitive_files parameter to indicate that it should use "default" order.

**Comment by ulfjack** [comment]
Is this still P1?

**Comment by endobson** [comment]
Which commit fixed this, it still looks broken on master? (Assuming I found the right part that maps to that generated documentation).

https://github.com/bazelbuild/bazel/blob/4ca8b8f98f043ef624b2d1605d658aa39c2c63dc/src/main/java/com/google/devtools/build/lib/rules/SkylarkRuleImplementationFunctions.java#L497

**Comment by brandjon** [comment]
Sorry, it's committed internally and should appear on github shortly.

**Comment by brandjon** [comment]
Fixed by e5d95fb.
